api/src/main/java/org/eclipse/microprofile/concurrent/ThreadContextBuilder.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.eclipse.microprofile.concurrent;
+
+import org.eclipse.microprofile.concurrent.spi.ConcurrencyProvider;
+
+/**
+ * <p>Builder for <code>ThreadContext</code> instances.</p>
+ *
+ * <p>Example usage:</p>
+ * <pre><code> ThreadContext threadContext = ThreadContextBuilder.create()
+ *                                                   .propagate(ThreadContext.APPLICATION, ThreadContext.SECURITY)
+ *                                                   .unchanged(ThreadContext.TRANSACTION)
+ *                                                   .build();
+ * ...
+ * </code></pre>
+ */
+public interface ThreadContextBuilder {
+    /**
+     * <p>Builds a new <code>ThreadContext</code> instance with the
+     * configuration that this builder represents as of the point in time when
+     * this method is invoked.</p>
+     *
+     * <p>After <code>build</code> is invoked, the builder instance retains its
+     * configuration and may be further updated to represent different
+     * configurations and build additional <code>ThreadContext</code>
+     * instances.</p>
+     *
+     * @return new instance of <code>ThreadContext</code>.
+     * @throws IllegalArgumentException if the same thread context type is
+     *         present in, or inferred by, both the {@link #propagate} set
+     *         and the {@link #unchanged} set.
+     */
+    ThreadContext build();
+
+    /**
+     * Creates a new <code>ThreadContextBuilder</code> instance.
+     *
+     * @return a new <code>ThreadContextBuilder</code> instance.
+     */
+    public static ThreadContextBuilder create() {
```

Comment:
I've updated it to rename the create method to instance on both builders.  However, please hold off on merging this pull because it still needs discussion.  I don't know if this will actually solve the issues raised by Stephane and I'm thinking about a different solution that might be better, in which case we wouldn't necessarily require these builders at all.